### ISSUE TITLE
Support <div (keyup.enter)="callback">

### ISSUE THREAD

**Issue by mhevery** [angular/angular#523]
Depends on #439 

It would be nice to create pseudo events on keys. So `keyup` and `keydown` would further be extended with: `space`, `enter`, `shift.enter`, `key_a`, and so on. It would need to understand, named keys, letters, modifiers.

The implementation can be done through the event manager (see #439)

**Comment by vicb** [comment]
I have started to investigate this.

- `keydown` and `keyup` have a `event.keyCode`which identify the physical key: the keyCode for 'a' and 'A' is the same (65). However "1" and "numpad_1" have different key codes.

- `keypress` events are only generated for printable characters and have a `event.charCode`, ie 65 for 'A' but 97 for 'a'. "1" and "numpad_1" has the same char code.

ref:
- https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent
- https://github.com/google/closure-library/tree/master/closure/goog/events -> key*.js

**Comment by MaartenVerbaarschot** [comment]
I like this idea.

@vicb You probably saw this already, but wanted to mention it anyway: You could design based on KeyboardEvent.key instead of keyCode since you are targeting future browsers with Angular 2.x: https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent.key

Then Polyfill the keyboard events using keyCode, charCode, etc. One available polyfill that I could find: https://github.com/termi/DOM-Keyboard-Event-Level-3-polyfill Haven't used it, so not sure if it's any good. No commits since May 2013. But you guys would probably write your own anyway? :p

**Comment by MaartenVerbaarschot** [comment]
I've got another idea that I would like to share, since you're investigating. This probably belongs in a separate issue but it is a scenario particularly relevant with keyboard events, so I'll start here. 

I've built apps (Smart TV prototypes) where I wanted to use the scope/context to maintain the life cycle of a key binding, but not bind it to the DOM element that the attribute is used on (I needed it on the document so the handler would trigger regardless of the focused element position). So, the way I approach this in Angular 1.x is write my own directives to bind everything to the document, and use the scope to create/destroy the bindings at appropriate moments.

When I translate my approach to 2.x syntax, it could look like this:
(Not sure if you're still using the `|` character for these situations. I intend it to be the same syntax as for referencing an item in NgRepeat.)

```html
<modal #modal (^keydown.esc|document)="modal.close()">
  <dropdown #dropdown (^keydown.esc|document)="dropdown.close()"></dropdown>
</modal>
```

In addition, in the case of having multiple bindings with the same key, I often needed a way to distinguish between scenarios where multiple handlers should be called, and scenarios where only the one with the highest priority should be called (in my case the scope that was created last, usually had the highest priority).

In the modal+dropdown example, it would enable you to specify that when pressing `esc`, only the dropdown is closed, and when pressing it again, the modal is closed. In my projects I did this by returning false in the expression, simply because I haven't thought of anything better yet. It has some resemblance to `stopPropagation()`, maybe there's inspiration to take from that in designing a better syntax.

```html
<modal #modal (^keydown.esc|document)="modal.close()">
  <dropdown #dropdown (^keydown.esc|document)="dropdown.close(); return false;"></dropdown>
</modal>
```

For me, these would be very nice additions to keyboard handling that I would use often. It helped me to keep the controllers very clean, and easily control key binding life cycles even when they're all on the document.

**Comment by teterovic** [comment]
Would be nice for mouseup.right
event.button=2

**Comment by thw0rted** [comment]
If anybody is still watching this issue, please take a look at #18870. This syntax isn't really documented and I'd like to know how to get it working with other keys by `event.key`.  As mentioned upthread, you can use the printable representation of the key if it's a number or letter, but many keys aren't valid as part of a DOM attribute or directive name (brackets, pipe, etc).
